File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Scanner;

public class Main {

    private static final int[] dx = {1, -1, 0, 0};

    private static final int[] dy = {0, 0, 1, -1};

    public static void main(String[] args) {
        try (Scanner scanner = new Scanner(System.in)) {
            int mapSize = scanner.nextInt(); // 지도 크기

            int[][] map = new int[mapSize][mapSize]; // 지도 선언
            boolean[][] visited = new boolean[mapSize][mapSize]; // 방문 여부

            for (int i = 0; i < mapSize; i++) {
                String input = scanner.next();
                for (int j = 0; j < mapSize; j++) {
                    map[i][j] = Character.getNumericValue(input.charAt(j));
                }
            }

            int homeCount = 0;
            List<Integer> homeSizes = new ArrayList<>();

            for (int i = 0; i < mapSize; i++) {
                for (int j = 0; j < mapSize; j++) {
                    if (!visited[i][j] && map[i][j] == 1) {
                        int size = dfs(map, visited, j, i);
                        homeSizes.add(size);
                        homeCount++;
                    }
                }
            }

            System.out.println(homeCount);
            Collections.sort(homeSizes);
            for (int size : homeSizes) {
                System.out.println(size);
            }
        }
    }

    private static int dfs(int[][] map, boolean[][] visited, int x, int y) {
        int count = 1;
        visited[y][x] = true;

        for (int direction = 0; direction < 4; direction++) {
            int nx = x + dx[direction];
            int ny = y + dy[direction];

            if (nx >= 0 && ny >= 0 && nx < map.length && ny < map.length && map[ny][nx] == 1 && !visited[ny][nx]) {
                count += dfs(map, visited, nx, ny);
            }
        }
        return count;
    }
}

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.kt
fun main() {
    val n = readln().toInt()

    val graph = Array(n) { IntArray(n) }
    val visited = Array(n) { BooleanArray(n) }

    for (i in 0..<n) {
        val input = readln().map { it.toString().toInt() }.toIntArray()
        for (j in 0..<n) {
            graph[i][j] = input[j]
        }
    }

    val result = ArrayList<Int>()

    for (i in 0..<n) {
        for (j in 0..<n) {
            if (graph[i][j] == 1 && !visited[i][j]) {
                val sum = 0
                result.add(dfs(n, j, i, graph, visited, sum))
            }
        }
    }

    result.sort()
    println(result.size)
    result.map {
        println(it)
    }
}

val dx = intArrayOf(1, 0, -1, 0)
val dy = intArrayOf(0, 1, 0, -1)

private fun dfs(
    n: Int,
    x: Int,
    y: Int,
    graph: Array<IntArray>,
    visited: Array<BooleanArray>,
    sum: Int
): Int {
    var count = 1
    visited[y][x] = true

    for (i in 0..3) {
        val dxx = x + dx[i]
        val dyy = y + dy[i]

        if (dxx >= 0 && dyy >= 0 && dxx < n && dyy < n && !visited[dyy][dxx] && graph[dyy][dxx] == 1) {
            count += dfs(n, dxx, dyy, graph, visited, sum)
        }
    }

    return count
}
